umbraco/Umbraco-CMS
# Performance improvement: Reusable data editors

### Prerequisites


### Description

This re-introduces the performance improvements we were forced to roll back in https://github.com/umbraco/Umbraco-CMS/pull/10938

This version makes it opt-in for property editors to reuse their data value editors. Opt-in is done by setting `ValueEditorIsReusable` in `DataEditorAttribute` to `true`. To be on the safe side, the property value is `false` by default. 

All core property editors except Block List and Nested Content opt in for data value editor reuse.

`IEditorConfigurationParser` is consumed by pretty much every single property editor in the core, among other things to create data value editors with configuration. I have opted to change its registration in the DI to singleton, as:
 
1. This is the case for all other services consumed by property editors, and
2. The implementation `EditorConfigurationParser` holds no state.

### Testing this PR

1. Create some content, save and reload - verify that property values are persisted as expected.
2. Create a Block List editor with one or more element types, and:
     - Create some content with this editor.
     - Go to settings and change one of the element types (i.e. add a new property).
     - Go back to the previously created content - verify that the element type change takes effect in the Block List elements.
     - Edit an element of the changed element type, save and reload - verify that the changed element type data is persisted.
